Add missing info into Add new resource through TF doc (#899)
GoogleCloudPlatform/k8s-config-connector: README.ConfigureResourceReferences.md

    the other GCP resource is needed when creating the target resource, then we
    know the field is likely a reference field. Note down the field name and the
    other GCP resource name.
1.  In the TF sample for the target resource, if the value of a field is an
    attribute reference of another TF resource, then we know the field is likely
    a reference field. Note down the field name and the other TF resource type
    name. E.g. the `key_ring` fields mentioned in the
    [cmek sample for PubSubTopic](https://registry.terraform.io/providers/hashicorp/google-beta/latest/docs/resources/pubsub_topic#example-usage---pubsub-topic-cmek).
1.  Be aware of field name mentioning: `name`, `id`, `link`, `email`,
    `service_account`, resource names under the same API of the target resource,
    or field description mentioning `refer`, `GCP`/`google`/`cloud` resources,
    `url`, `uri`, etc. Once you see these keywords, it's very likely that this
    field references to gcp resource. Double-check the GCP tutorial and/or TF
    documentation to confirm.

Figure out the referenced GCP resource through reading the field description. If

> you are still unsure, contact the API team for more information. Once it's

Reviewer: Shall we just skip mentioning this? An OSS contributor may not be able to contact the API team.

Author: Removed

determined, note down the field name and the referenced GCP resource name.

## Identify whether the referenced resource is TF-based or DCL-based

### Turn GCP resource name or TF type name into KCC kind name

Before moving on, you need to turn the referenced TF type/GCP resource names
you've noted down into KCC kind names. Ask the KCC team for help if you still

> can't figure them out with the following instructions.

Reviewer: "Ask the KCC team" seems not very actionable in the context of OSS. We can say file a GitHub question in this repo instead.

1.  If you know the TF type name, you can do project-scoped search in your local
    repository with the TF type name. If you find a hit in a service mapping
    yaml file (under either the `config/servicemappings/` folder or the
    `scripts/resource-autogen/generated/servicemappings/`), the corresponding
    `kind` field in the same resource config should be the KCC kind.

1.  If you know the GCP resource name, you can either use the search widget to
    find the corresponding TF type in
    [Terraform doc](https://registry.terraform.io/providers/hashicorp/google-beta/latest/docs)
    with the GCP resource name, then figure out the KCC kind name following the
    step above; or you can use the search widget to find the corresponding KCC
    kind directly in
    [KCC reference doc](https://cloud.google.com/config-connector/docs/reference/overview).

> 1.  If you can't find a KCC kind with the steps above, it's possible that the

Reviewer: Similarly instead of saying "Reach out to the KCC team", let's just ask the contributor to raise a question as a GitHub issue. Or we simply remove this sentence.

    referenced resource is not supported in KCC. Reach out to the KCC team is
    you are unsure. If you've determined that the KCC kind is not supported in
    KCC, skip the next couple of sections and go to
    [Configure reference resource in the service mappings](#configure-reference-in-the-service-mappings).

> ### Determine whether the referenced resource is TF-based or DCL-based

Reviewer 2: not sure what DCL stands for (tried to Google it but Disney Cruise Lines has the SEO here 😅 ) but we should probably expand this acronym, too, alongside Terraform

Author: It's Declarative Client Library haha

Based on the KCC kind names of the references you figured above, follow the

> steps to identify the referenced resource is DCL or TF based.

Reviewer 2: ```suggestion
steps to identify if the referenced resource is DCL or TF-based:
```

1.  Search for the KCC kind name of the referenced resource on the
    [crds folder](https://github.com/GoogleCloudPlatform/k8s-config-connector/tree/master/crds).

> 1.  Open the CRD file containing the KCC kind name, search for `dcl2crd` and

Reviewer 2: ```suggestion
1.  Open the CRD file containing the KCC kind name. Search for `dcl2crd` and
```

>     `tf2crd` labels (only one should exist for a particular CRD).

Reviewer 2: ```suggestion
    `tf2crd` labels. There should be only one label for each CRD.
```

>     1.  If `cnrm.cloud.google.com/dcl2crd: "true"` is found, then it is a

Reviewer 2: ```suggestion
    1.  If the resource has the `cnrm.cloud.google.com/dcl2crd: "true"` label, it's a
```

        DCL-based resource.

>     1.  If `cnrm.cloud.google.com/tf2crd: "true"` is found, then it is a

Reviewer 2: ```suggestion
    1.  If the resource has the `cnrm.cloud.google.com/tf2crd: "true"` label, it's a
```

        TF-based resource.

> ## Identify whether the reference field is a string field or a field of other types (list, etc)

Reviewer 2: could we shorten this title to something like "Identify the reference field type" and then expand the description to something like "Next, you identify the type of the reference field. In most cases, the reference field is a string field, but could be another type like a list. Check the Terraform schema of the target resource..."

Check the TF schema of the target resource in the TF provider to determine the
type of the reference field. The URL of file is in the format of
`https://github.com/hashicorp/terraform-provider-google-beta/blob/main/google-beta/services/[servicename]/resource_[tf_type_name_without_google_prefix].go`.
You can find a `Type` field for each argument, and two most common types are
`schema.TypeString` and `schema.TypeList`.

1.  If the field is a string, follow
    [Configure reference resource in the service mappings](#configure-reference-in-the-service-mappings)
    to continue configuring the resource reference.
1.  If the field is a list:

    1.  If the field is not required, mark it as an ignored field following
        [this example](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/b746248cd5a9b30669380513de8fdc6b4c43018d/config/servicemappings/cloudbuild.yaml#L204).
    1.  If the field is required, reach out to the KCC team for further

>         discussion.

Reviewer: let's suggest to raise a GitHub question. For our team members we can always discuss internally.

> 1.  If the field is neither a string nor a list, reach out to the KCC team for

Reviewer 2: ```suggestion
1.  If the field isn't a string or a list, reach out to the KCC team for
```

    further discussion.

## Configure reference resource in the service mappings

> Based on the information identified in above sections, configure the service

Reviewer 2: ```suggestion
Based on the information identified in the above sections, configure the service
```

mapping file by providing values for the required fields. Fields can be found in
[service mapping types](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/b746248cd5a9b30669380513de8fdc6b4c43018d/pkg/apis/core/v1alpha1/servicemapping_types.go#L242)

Here is an example configuration:
```
- tfField: disk.source_snapshot_encryption_key.kms_key_service_account
  key: kmsKeyServiceAccountRef
  description: |-
    The service account being used for the encryption request for the
    given KMS key. If absent, the Compute Engine default service account
    is used.
  gvk:
    kind: IAMServiceAccount
    version: v1beta1
    group: iam.cnrm.cloud.google.com
  targetField: email
```

> 1.  `tfField`: TF field which has reference to other GCP resources. In above

Reviewer 2: ```suggestion
1.  `tfField`: TF field which references a GCP resource. In the above
```

        further discussion.

1.  If the field isn't a string or a list, create a Github question in this repo
    for further discussion.

## Configure reference resource in the service mappings

Based on the information identified in the above sections, configure the service
mapping file by providing values for the required fields. Fields can be found in
[service mapping types](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/b746248cd5a9b30669380513de8fdc6b4c43018d/pkg/apis/core/v1alpha1/servicemapping_types.go#L242)

Here is an example configuration:
```
- tfField: disk.source_snapshot_encryption_key.kms_key_service_account
  key: kmsKeyServiceAccountRef
  description: |-
    The service account being used for the encryption request for the
    given KMS key. If absent, the Compute Engine default service account
    is used.
  gvk:
    kind: IAMServiceAccount
    version: v1beta1
    group: iam.cnrm.cloud.google.com
  targetField: email
```

1.  `tfField`: TF field which references a GCP resource. In the above
    example, it's `disk.source_snapshot_encryption_key.kms_key_service_account`.

>     Note that it is snack case, and it should be a path if it has parent fields.

Reviewer: snake not snack. :-)
